src/geometry.ts

```diff
@@ -0,0 +1,304 @@
+	const borderBox = style.boxSizing && style.boxSizing === 'border-box';
+
+	if (rect.height) {
+		let height = 0;
+		if (borderBox) {
+			height = rect.height;
+		}
+		else {
+			height = rect.height
+				- pxToNumber(style.paddingTop)
+				- pxToNumber(style.paddingBottom)
+				- pxToNumber(style.borderTopWidth)
+				- pxToNumber(style.borderBottomWidth);
+		}
+		element.style.height = numberToPx(height);
+	}
+
+	if (rect.width) {
+		let width = 0;
+		if (borderBox) {
+			width = rect.width;
+		}
+		else {
+			width = rect.width
+				- pxToNumber(style.paddingLeft)
+				- pxToNumber(style.paddingRight)
+				- pxToNumber(style.borderLeftWidth)
+				- pxToNumber(style.borderRightWidth);
+		}
+		element.style.width = numberToPx(width);
+	}
+
+	if (rect.top) {
+		element.style.top = numberToPx(rect.top - pxToNumber(style.marginTop));
+	}
+
+	if (rect.left) {
+		element.style.left = numberToPx(rect.left - pxToNumber(style.marginLeft));
+	}
+}
+
+/**
+ * Sets the dimensions of an Element's content box.
+ * @param element The Element to modify
+ * @param rect An object containing optional number properties for height, left, top, and width
+ */
+export function setContentBox(element: HTMLElement, rect: SettableClientRect): void {
+	const style = getComputedStyle(element);
+	const borderBox = style.boxSizing && style.boxSizing === 'border-box';
+
+	if (rect.height) {
+		let height = 0;
+		if (borderBox) {
+			height = rect.height
+			+ pxToNumber(style.paddingTop)
+			+ pxToNumber(style.paddingBottom)
+			+ pxToNumber(style.borderTopWidth)
+			+ pxToNumber(style.borderBottomWidth);
+			
+		}
+		else {
+			height = rect.height;
+		}
+		element.style.height = numberToPx(height);
+	}
+
+	if (rect.width) {
+		let width = 0;
+		if (borderBox) {
+			width = rect.width
+			+ pxToNumber(style.paddingLeft)
+			+ pxToNumber(style.paddingRight)
+			+ pxToNumber(style.borderLeftWidth)
+			+ pxToNumber(style.borderRightWidth);
+		}
+		else {
+			width = rect.width;
+		}
+		element.style.width = numberToPx(width);
+	}
+
+	if (rect.top) {
+		element.style.top = numberToPx(rect.top
+			- pxToNumber(style.marginTop)
+			- pxToNumber(style.borderTopWidth)
+			- pxToNumber(style.paddingTop));
+	}
+
+	if (rect.left) {
+		element.style.left = numberToPx(rect.left
+			- pxToNumber(style.marginLeft)
+			- pxToNumber(style.borderLeftWidth)
+			- pxToNumber(style.paddingLeft));
+	}
+}
+
+/**
+ * Sets the dimensions of an Element's margin box.
+ * @param element The Element to modify
+ * @param rect An object containing optional number properties for height, left, top, and width
+ */
+export function setMarginBox(element: HTMLElement, rect: SettableClientRect): void {
+	const style = getComputedStyle(element);
+	const borderBox = style.boxSizing && style.boxSizing === 'border-box';
+
+	if (rect.height) {
+		let height = 0;
+		if (borderBox) {
+			height = rect.height
+				- pxToNumber(style.marginTop)
+				- pxToNumber(style.marginBottom);
+		}
+		else {
```

Comment:
This can be simplified to:

```ts
let height = rect.height - pxToNumber(style.marginTop) - pxToNumber(style.marginBottom);
if (!borderBox) {
    height -= ...
}
```

(where the part in the if would have just the padding/border subtractions; same goes for width below)